### ISSUE TITLE
Pre and post hooks for assemble/run/etc scripts

### ISSUE THREAD

**Issue by php-coder** [openshift/source-to-image#573]
@thoraxe @GrahamDumpleton @thesteve0 @bparees,

To improve user experience, when users are trying to add only a pre/post actions to standard `assemble` and `run` scripts, I'm collecting use-cases of that. I would like to ask you to share real use-cases or your thoughts whether this feature would be useful or it wouldn't.

Thank you!

Related links:
- https://github.com/openshift/source-to-image/issues/154
- https://trello.com/c/WpdJNDz7/524-5-as-a-sti-user-i-d-like-to-extend-the-default-scripts-evg

**Comment by php-coder** [comment]
The first use-case that @thoraxe describes in his [blog post](http://www.erikjacobs.com/2016/07/21/node-js-gulp-and-openshift-3-custom-assemble-script-ftw/), is to do a post-processing of the templates with gulp inject. As far I understand, it even could cover all the actions with gulp, because our assemble script knows nothing about gulp and build-in assemble can be used only for installing gulp.

**Comment by thoraxe** [comment]
I kinda miss the old V2 hooks system.

It would be nice if we had the option to have .sti/bin/assemble-{pre,post} if you want to use the existing script and then just run things before/after. But being that you can simply reference the script itself (https://trello.com/c/WpdJNDz7/524-5-as-a-sti-user-i-d-like-to-extend-the-default-scripts-evg#comment-564482fecb0867efa9ac6129) which I did not realize in the original #154, I'm kinda thinking about just closing #154.

**Comment by bparees** [comment]
@thoraxe introduce pre-assemble/post-assemble is part of what we're discussing which is why we want use cases to understand if introducing those actually solves enough of the problem to be useful (as opposed to use cases where you need to modify something in the middle of the assemble script execution).

and then the other debate is, as you've noted/implied, whether it's worth introducing those minimal hook points given that users can already effectively accomplish the same thing.

**Comment by GrahamDumpleton** [comment]
The action hook types you need and why as are as follows.

In the ``assemble`` phase you need:

* ``pre_build`` - Executed prior to building application artefacts. This is needed to handle the case where a user needs to install a library or other package from source code which is not part of system build but is required to install application dependencies which are built/installed automatically, such as from Python package index (PyPi). Because system directories are not writable, the user would need to install any library or package into an area the user has access to. Where this may be is potentially going to differ based on S2I builder used due to fact that S2I builders may not use a consistent directory layout.
* ``build_env`` - This is not executed separately, but would be executed inline using ``source`` in the ``assemble`` script.  It would be run such that any variables set in the hook would be automatically exported and user doesn't need to use ``export`` themselves. This is needed to allow a user to set environment variables dynamically, which are required for the latter build steps to install application dependencies or build other application artefacts. This would be required where ``pre_build`` was used to install a library or package and environment variables had to be set so that when installing application dependencies, it knows where to find the header files/libraries or programs, installed by the ``pre_build`` script. It might also be used to set up additional environment variables dynamically based on static values from the Docker image. The static nature of ``.s2i/environment`` is too limiting and why this way of dynamically overriding environment variables for later build steps is needed.
* ``build`` - This is executed after application dependencies or other application artefacts have been built. This is needed where a build action is dependent on having application dependencies installed first.

In the ``run`` phase you need:

* ``deploy_env`` - This is not executed separately, but would be executed inline using ``source`` in the ``run`` script. It would be run such that any variables set in the hook would be automatically exported and user doesn't need to use ``export`` themselves. This is needed to allow a user to set environment variables dynamically, which are required for the latter deployment steps or when starting up the application. This would be required where the environment variables need to be set dynamically based on values derived from a package installed as part of the build. It might also be used to set up additional environment variables dynamically based on static values from the Docker image. The static nature of ``.s2i/environment`` is too limiting and why this way of dynamically overriding environment variables for later build deployment is needed.
* ``deploy`` - This is executed just before the ``run`` script starts up the application. This would be required where last minute steps to prepare data for the application is required, or where it may be necessary to run additional tasks as background processes.

Because setting of environment variables is allowed dynamically, and those same environment variables from ``deploy_env`` could be required in any interactive shell environment created using ``oc rsh``, or script executed using ``oc exec``, then ``deploy_env`` would need to be run inline with any script used to enable SCL packages via ``ENV``, ``BASH_ENV`` or ``PROMPT_COMMAND`` hack.

One final twist that might necessitate an additional ``deploy_cfg`` action hook is where an application config file required before starting the application, needs to incorporate values from environment variables, static or dynamic. Editing on the fly of such configuration files to incorporate values from environment variables could be done from ``deploy_env`` but if that were done, it would be on the user to use a marker file to ensure that they don't run the action more than once, given that ``deploy_env`` could technically be executed more than once in the life of the container. Thus ``deploy_cfg`` could be implemented as a separate concept, with the ``run`` script, plus SCL enable script, managing the marker files itself to ensure not run more than once, thus meaning user doesn't need to worry about it.

* ``deploy_cfg`` - This is executed after ``deploy_env`` and before ``deploy``. It is required where it is necessary to generate application configuration files which incorporate values from environment variables. This must only be run once. It needs to also be run, if not already run, as a side effect of scripts run using ``oc exec`` or when a shell is created using ``oc rsh``. This is necessary for cases of where using lifecycle hooks and it is only a script that is run, and not the main ``run`` script.

You can see an example implementation based around an existing S2I builder for the above, with exception of ``deploy_cfg`` at:

* https://gitlab.com/osevg/s2i-action-hooks

It doesn't implement the ``deploy_cfg`` hook at this point due to uncertainties around whether S2I builders follow a consistent directory structure and thus where one can safely write a marker file to track whether the hook has already been run.

You can see a separate example where a config hook is used in:

* https://github.com/GrahamDumpleton/warpdrive/blob/master/warpdrive/etc/warpdrive-start#L27

This latter example doesn't currently have equivalent of ``build_env`` and ``deploy_env`` hooks as it was trying to do it with a single hook script for setting environment for both build and deploy, but that has been found to be less than ideal given that it would be run twice and user would need to work out whether build or deployment and set environment variables selectively. Thus two separate scripts seen as better. So you don't get confused, this latter example is a generic system that can operate with Docker alone and doesn't require S2I. It works with S2I through a small shim wrapper to many from S2I script names to the way the system does things.

Since availability of any changes to accomodate action hooks is likely to be many months away, and because I have had to explain how to do this sort of stuff to multiple people over the past few weeks, I was aiming to write a blog post about how to implement it using the ``s2i-action-hooks`` repository as sample.

Do be aware that since that the ``s2i-actions-hooks`` sample is having to wrap existing ``assemble`` script, it needs to accomodate for the fact that source code would not have been copied into place yet. As a result ``pre_build`` hook needs to use files out of ``/tmp/src``if necessary. Where integrated properly into existing ``assemble`` script of an S2I builder, then the location of where hook scripts run can be better located, such as ``pre_build`` being after source code moved into correct location. Exact placement may vary based on specific S2I builders and the language used. Some S2I builders might need additional hooks, but the above would be a minimal set which covers bulk of potential use cases.

**Comment by jorgemoralespou** [comment]
@GrahamDumpleton I agree with your explanation and I think this cover most of the cases.
Also I think that extended builds (or if something in the future replaces this) could/should split the build/assemble and the run phases, so making this more clear whenever crafting the scripts would greatly help.

**Comment by php-coder** [comment]
@GrahamDumpleton Thank you for sharing with us such comprehensive explanation!

>    pre_build - Executed prior to building application artefacts. This is needed to handle the case where a user needs to install a library or other package from source code which is not part of system build but is required to install application dependencies which are built/installed automatically, such as from Python package index (PyPi). Because system directories are not writable, the user would need to install any library or package into an area the user has access to. Where this may be is potentially going to differ based on S2I builder used due to fact that S2I builders may not use a consistent directory layout.

1) `pre_build` sounds logical and natural but because S2I script for building is called `assemble` then I would rather call the pre hook as `pre_assemble`

2) if to build an application user needs to **install a library to the system directories** then clean and right way will be to create a custom builder that will have these libraries already installed. If I have to use a gem that requires system libraries, I won't download/install GCC and library's sources, compile them and after that install my gem, right? Unlike application's sources we can use the same version of the system library for all of the builds. In other words, this is a non-standard task and we shouldn't solve such tasks on the level of the `assemble` script.

(And even if user tries to do so, he always has writable /tmp directory.)

>    build_env - This is not executed separately, but would be executed inline using source in the assemble script. It would be run such that any variables set in the hook would be automatically exported and user doesn't need to use export themselves. This is needed to allow a user to set environment variables dynamically, which are required for the latter build steps to install application dependencies or build other application artefacts. This would be required where pre_build was used to install a library or package and environment variables had to be set so that when installing application dependencies, it knows where to find the header files/libraries or programs, installed by the pre_build script.

It is the consequence of choosing the wrong way. Preparation of the system for building and preparation of the sources are on the different levels. I'm trying to focus on solving application level tasks but this one should be fixed on the lower level by creating a custom builder.

> It might also be used to set up additional environment variables dynamically based on static values from the Docker image. The static nature of .s2i/environment is too limiting and why this way of dynamically overriding environment variables for later build steps is needed.

I can't imagine real use-case for that, could you give me an example?

BTW maybe we can use the Ansible-way: if `.s2i/environment` is executable then S2I will `source` it. This could solve the problem with static nature of `.s2i/environment` file.

>    build - This is executed after application dependencies or other application artefacts have been built. This is needed where a build action is dependent on having application dependencies installed first.

When this should be executed — also before the `assemble` script?

>    deploy_env - This is not executed separately, but would be executed inline using source in the run script. It would be run such that any variables set in the hook would be automatically exported and user doesn't need to use export themselves. This is needed to allow a user to set environment variables dynamically, which are required for the latter deployment steps or when starting up the application. This would be required where the environment variables need to be set dynamically based on values derived from a package installed as part of the build. It might also be used to set up additional environment variables dynamically based on static values from the Docker image. The static nature of .s2i/environment is too limiting and why this way of dynamically overriding environment variables for later build deployment is needed.

Why are you named it `deploy` when it relates to `run` script? Deploy could puzzle users.

I see here that we're trying to solve another issue: sometimes we want to have different environments in the builder and runtime containers. I think that this should be solved separately.

Also it can be hard to "execute inline using source in the run script" because it requires modification of the `run` script. S2I doesn't manage and modify this script, it just executes it.

>    deploy - This is executed just before the run script starts up the application. This would be required where last minute steps to prepare data for the application is required, or where it may be necessary to run additional tasks as background processes.

>Because setting of environment variables is allowed dynamically, and those same environment variables from deploy_env could be required in any interactive shell environment created using oc rsh, or script executed using oc exec, then deploy_env would need to be run inline with any script used to enable SCL packages via ENV, BASH_ENV or PROMPT_COMMAND hack.

Sounds complicated and looks like it does the same as put this script into `/etc/profile.d`

>One final twist that might necessitate an additional deploy_cfg action hook is where an application config file required before starting the application, needs to incorporate values from environment variables, static or dynamic. Editing on the fly of such configuration files to incorporate values from environment variables could be done from deploy_env but if that were done, it would be on the user to use a marker file to ensure that they don't run the action more than once, given that deploy_env could technically be executed more than once in the life of the container. Thus deploy_cfg could be implemented as a separate concept, with the run script, plus SCL enable script, managing the marker files itself to ensure not run more than once, thus meaning user doesn't need to worry about it.

This is the case when user have to provide his own `run` script or even write a wrapper around his application.

>Do be aware that since that the s2i-actions-hooks sample is having to wrap existing assemble script, it needs to accomodate for the fact that source code would not have been copied into place yet. As a result pre_build hook needs to use files out of /tmp/src if necessary

This sounds like a bug or at least something that can be improved.

Thank you for your time again, your feedback gave us a lot of information to think of.

**Comment by bparees** [comment]
> if to build an application user needs to install a library to the system directories then clean and right way will be to create a custom builder that will have these libraries already installed. 

Trying to do that in a hook script is also likely to run into permission problems since the scripts will not be run as root, so that's another good reason to customize the image instead.

**Comment by GrahamDumpleton** [comment]
The naming of the hook scripts using 'build' and 'deploy' is based on OpenShift 2 because it had a good naming convention and many users will be used to that. It only confuses people by completely ignoring the past and creating a completely different way of doing things.

I am not talking just about installing additional packages into system directories. Not everyone wants to go off and create a new custom S2I builder to install as a library what could be an application specific code library that can't be handled using the language packaging mechanism, that only makes their workflow even more complicated as they can't do everything in the build and instead then need to worry about building a new base image all the time as well if the library is being changed.

Various builders already have a compiler as part of the base image. Just because some builders don't, and because you haven't yourself encountered situations where building a custom library is necessary, doesn't mean one should take a very narrow view of what functionality an action hooks mechanism should provide. Both the pre_build and then subsequent build_env can be important in some cases.

You can't use dynamic evaluation of ``.s2i/environment`` as ``s2i`` embeds those variables in the Docker image itself. They aren't dealt with by the ``assemble`` or ``run`` scripts. It is needed to more easily allow users to copy environment variables into differently named environment variables, or multiple environment variables as dictated by application code or dependent packages where changing that code is more trouble than its worth or not possible. Ability to set dynamically also allows defaults to more easily be set where they aren't set from outside.

The ``build`` phase would be executed after the specific language S2I builder had installed packages using its system, such as ``pip`` in Python. The separate example repo I linked shows the relationship. It has to be after as this is where users can run steps which need those language packages installed first.

If you think the ENV, BASH_ENV or PROMPT_COMMAND hack is complicated, ask the developers why it was done that way in the first place for some S2I builders. They must have had there reasons. Personally I don't like it either and prefer a helper script inside of the image which is executed and given as arguments the real command to run. That way the helper script can set up the environment without having to resort to those hacks. There is an entrypoint script in some images which sort of provides that, but it is not part of what you can officially use, is not in PATH and could even be in different locations based on which builder it is since no guarantees about it.

As to deploy_cfg, the user can't just provide their own run script which wraps the running of the application. This is because the S2I builders don't provide that capability. They would have to write their own S2I builder. This is why the action hooks mechanism is needed in the first place, to allow such overrides. Wrapping the startup of the application also doesn't help with situation where that same step of generating configuration from environment variables needs to be done when running management commands such as in a lifecycle hook. In that case the target application is never even run, so having a wrapper around it doesn't help. Having to wrap every single management command you might need to run separately in a container is just extra work where it could all be handled in one way.

And finally, needing to reference stuff out of /tmp/src is not a bug, it is part of how S2I works to get the source code into an image. One could avoid it by integrating pre_build into the assemble script at the point after code is moved into place based on how S2I builder works, but I have come to the conclusion that it is better before with some working off /tmp/src at that point as allows for application to better override certain things about how code is moved around.

**Comment by bparees** [comment]
> The naming of the hook scripts using 'build' and 'deploy' is based on OpenShift 2 because it had a good naming convention and many users will be used to that. It only confuses people by completely ignoring the past and creating a completely different way of doing things.

When technology changes, names sometimes need to change to.  We don't call things "gears" and "cartridges" anymore either.  "run" most closely maps to what you are doing with the image because it corresponds to what happens when you "docker run" the image.  And we're not going to go back and rename s2i's assemble step to "build" so as @php-coder says, pre_assemble makes more sense for a set of users who are going to need to become accustomed to "assemble" as a terminology (and assemble makes more sense than build since it's assembling an image, not just building source code).

> Various builders already have a compiler as part of the base image. Just because some builders don't, and because you haven't yourself encountered situations where building a custom library is necessary, doesn't mean one should take a very narrow view of what functionality an action hooks mechanism should provide. Both the pre_build and then subsequent build_env can be important in some cases.

this is why we reached out to you, so you can provide examples of these requirements and help us understand if it's a "90% of people will need this" or "10% of people will need this".  These aren't easy problems to solve, we're not going to boil the ocean for 10% of users who would be better served to just replace the assemble script entirely.  That said, i'm not seeing what about your scenario isn't handle by a pre_assemble hook in this case, so hopefully we can reach an agreement that such a feature is a) useful and b) sufficient.


> You can't use dynamic evaluation of .s2i/environment as s2i embeds those variables in the Docker image itself. 

@php-coder was suggesting that there be two different modes.  Yes today the contents of the "environment" file get processed before we launch the assemble container and they are part of the container and thus included in what is committed.  But there could be an alternate mode in which instead of doing that, we simply "source" the environment file before running assemble.  That would be done inside the shell of the container, it would not alter the container config itself, and thus would not be part of the committed image.  I don't actually like overloading the meaning of "environment", i'd prefer a separate file, but the general concept remains the same.

> The build phase would be executed after the specific language S2I builder had installed packages using its system, such as pip in Python. 

there's basically no chance we're going to break assemble up into discrete sub-steps at this point with the ability to call hooks in between those steps.  Doing so would require either:

1) s2i image authors write assemble scripts that explicitly call out to hooks if they exist (too hard for authors to implement correctly, not backwards compatible, makes writing an assemble script significantly more complicated)
2) s2i explicitly breaks assemble up into multiple scripts ("assemble_deps","assemble_compile","assemble_package", etc).  Again this has backwards compatibility problems (even if we continue to tolerate an "assemble" script, someone using an image that only supplies assemble, while trying to leverage a "pre_assemble_compile" hook is not going to get a valid result) and introduces a higher burden on s2i script authors to break their assemble script up in logic ways (and in ways that may not even make any sense for their particular framework).

So unless we can demonstrate that this level of granularity is something 90% of users will need on a regular basis (which was the point of starting this discussion), I think we have to continue to punt this into the "override the assemble script either in your source repo or by creating your own custom s2i builder image" category, for cases that can't be handled by a pre_assemble or post_assemble hook.

v2 had a model closer to what you're describing and while it was very flexible, it seemed like:

1) most people didn't use hooks for much (at most a pre_start hook here and there)
2) very few people wrote new cartridges because the burden was too high to get those details correct

> If you think the ENV, BASH_ENV or PROMPT_COMMAND hack is complicated, ask the developers why it was done that way in the first place for some S2I builders. They must have had there reasons. 

it was done to ensure that no matter how you run the image or exec into a running image, the scl packages are enabled for you.  Just running a helper script via the entrypoint doesn't solve that, and i'm pretty sure /etc/profile.d isn't consistently processed depending on how entered the container.

> As to deploy_cfg, the user can't just provide their own run script which wraps the running of the application. This is because the S2I builders don't provide that capability. 

What do you mean?  you can replace the run script, or you can write a run script that, as its last action, calls the existing image run script.

**Comment by GrahamDumpleton** [comment]
The ENV, BASH_ENV or PROMPT_COMMAND does not work to always enable SCL packages. It will only work if you are running a shell script which then does what you need. It will not do what is required for something like:

```
oc exec xxx python ...
```

It will pick up the system Python and not the SCL Python.

The use of a helper script and saying to users that they should always use it when executing stuff in the container is more predictable.

```
oc exec xxx warpdrive exec python ...
```

The ``warpdrive exec`` command in this case always ensures that SCL is enabled, plus any environment variables set by hook scripts would always be set.

Wrapping the whole ``run`` script doesn't help where needing to do stuff from life cycle hooks. Have to replicate it for that as well. So am talking about in context of having one way which works in both cases, that is what isn't supported.

Anyway, lets call time on this. Perhaps should just be closed as not fix as can't see any progress ever being made on this.

**Comment by thoraxe** [comment]
I'm not sure how you can complain about picking up system Python versus SCL Python when that is the behavior whether or not you are in a Docker container or even doing anything with OpenShift.

"oc exec" will execute a command in the remote container. It's just wrapping "docker exec". If you "docker run registry..../python -it bash" and then do "python foo", you're going to get the system python.

In this particular case (oc exec) you are asking for OpenShift to make an assumption about what the user wants to do with the remote container where it actually doesn't know anything about the remote container.

The only way that this particular thing could work, in my understanding, is if "oc exec" actually looked at metadata on the container image that it was being run against, and, depending on that metadata, "wrapped" whatever you were trying to exec inside of an "scl enable".

In other words, there would need to be Docker/container metadata that was something like "openshift.exec.wrapper" and, if present, would be used... or something.

I'm also still not understanding the need to build a library that isn't in the s2i but that isn't "built" as part of the native language dependency processing. Ruby gems, for example, will get built if they are required but not present. This may involve complication. This "just works" as it stands already. Is Python not capable of that?

**Comment by php-coder** [comment]
To sum up: right now we didn't decide anything except to put this off for the future.

**Comment by bparees** [comment]
@php-coder I think we decided/established more than that:

We established that the only realistic thing we can do here is pre/post hooks and that the target users (in this case our evangelists) felt that those did not provide sufficient benefit over the existing ability to replace the entire script and/or write a new script that invokes the existing script, to justify investing in building the feature.

We also defined some additional use cases (setting env variables) which can be covered by the aforementioned capabilities (env variables defined in shell scripts won't be in the final image).

Lastly, we concluded it was important to document how to reference the existing scripts when writing new scripts (in lieu of providing the explicit pre_assemble/post_assemble hook points), and produced product documentation to that effect.

@sspeiche fyi, this was one of the 3.4 items, so you should be aware of the direction it took.

**Comment by GrahamDumpleton** [comment]
As this discussion is being pushed out to others who may not be aware of the bigger picture beyond this issue, I will fill in that bigger picture as to where I am coming from and why I see this issue as not meeting the requirements I at least have.

When you look at languages such as node.js and npm, or Java and tools like Maven, they have quite developed eco systems for handling the build and deployment of applications. They already include within those systems ways of hooking in custom actions as different points within the build and deployment process. For example for npm you have:

* http://www.marcusoft.net/2015/08/pre-and-post-hooks-for-npm-scripting.html

In the Python world the build tools are not that sophisticated, handling package installation and nothing else. You cannot use the package installation system to readily trigger any custom actions you may need to do around or as part of the installation process. There are also so many different ways of deploying Python web applications and there is not one single dominant method used. There is therefore no commonly accepted way of having custom actions run by the deployment server either.

The lack of easy customisation points in the Python S2I builder is thus going to be felt much more acutely than builders for other languages.

In trying to get hook points introduced into the Python S2I builder there was push back to say that it must be done consistently across all builders and thus why this issue against source-to-image has become the focus point for discussion.

Unfortunately in trying to address it for all S2I builders you are trying to come up with a simplified one size fits all strategy which ignores that specific languages and the ecosystems around them may have different requirements. Some of the comments challenging why something would even be required show that there is lack of exposure to the extra challenges that exist in Python due to the short comings of the package installer for Python, plus the diversity of both Python web application servers and Python web frameworks or web toolkits.

The desire to keep things simple and avoid providing deeper integration with a specific S2I builder means that for Python at least, the solution you seem to want may not meet the requirements, or could be just as unwieldy as having nothing and telling people to work out a way of doing it themselves. It may even be better to recommend they use a different Python S2I builder developed by the community which can better address what the Python community feels they need.

So in that context of me trying to develop something that better meets the Python communities requirements, this back and forth discussion is not productive and I am better off working with the community to develop their own solution. There are also with the S2I builder more series issues which still aren't being addressed which this discussion just seems to be taking focus away from. So maybe if this issue is dropped the more serious issues will be fixed.

**Comment by bparees** [comment]
@GrahamDumpleton thanks, the distinction of "this functionality needs to be provided in a python specific way" vs "this problem should be solved in a generic way that works for all s2i builders" is a valuable one and definitely a key point that I was not picking up on until your most recent explanation.

I am definitely not averse to making the s2i-python builder image's assemble logic more sophisticated so as to handle the scenarios you have in mind, if indeed we are convinced that such logic is not generically applicable to s2i as an ecosystem.  (Or we are convinced that there is no suitable way to bring such generic logic to s2i(ie a way that doesn't make writing s2i builders significantly more painful or fragmented) and less need for it in other builders, which I think might be closer to the situation here).

When we started this issue, we were operating on the impression that significant numbers of users were finding it necessary to override the default scripts.  I'm still not sure we've actually gotten data to confirm that one way or another, or to answer the secondary question of whether pre/post hooks are sufficient to prevent users from having to replace the entire script when they do need to override behavior, but it sounds like at least in the case of python, the answer is "yes it's typical to need to override and no pre/post hooks aren't sufficient".

If anyone else on this thread has data about their experiences w/ other images, we'd still be interested in that, but for now we're proceeding with the documentation approach over implementing new capabilities.

**Comment by dcortesf** [comment]
Hi folks, I'm writing from Santander Group, we have been working with OpenShift from a long time. From our point of view, we think S2I should support hooks or extension points which allow users to run corporative process at any time (qa, security, etc). If S2I doesn't support this feature, we need use Jenkins and it no sense to use S2I for only for build task.

**Comment by GrahamDumpleton** [comment]
@dcortesf You might be interested in:

* https://github.com/getwarped/action-hooks

It provides an example structure for doing it yourself within your source code repository.
